llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -2185,6 +2273,172 @@ void VPlanTransforms::handleUncountableEarlyExit(
   LatchExitingBranch->eraseFromParent();
 }
 
+/// This function tries to match following pattern to create
+/// VPExtendedReductionRecipe and clamp the \p Range if it is beneficial and
+/// valid. The created VPExtendedReductionRecipe will lower to concrete recipes
+/// before execution.
+///   reduce(ext(...)).
+static VPExtendedReductionRecipe *
+tryToMatchAndCreateExtendedReduction(VPReductionRecipe *Red, VPCostContext &Ctx,
+                                     VFRange &Range) {
+  using namespace VPlanPatternMatch;
+
+  Type *RedTy = Ctx.Types.inferScalarType(Red);
+  VPValue *VecOp = Red->getVecOp();
+  const RecurrenceDescriptor &RdxDesc = Red->getRecurrenceDescriptor();
```

Comment:
Does this need anything from RdxDesc other than the info retained in @lukel97 's https://github.com/llvm/llvm-project/pull/131300

Reply:
Yeah, the recurrence kind can be obtained from just the opcode so I think this is fine